File: ios/MapConductorMapLibreViewManager.m
#import "MapConductorMapLibreViewManager.h"
#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>

@interface MCMapLibreReactNativeView : UIView
@property(nonatomic, copy) void (^eventHandler)(NSString *, NSDictionary *);
- (void)setCameraPosition:(NSDictionary *)payload;
- (void)setMapDesignType:(NSString *)value;
- (void)setMarkerTilingOptions:(NSDictionary *)payload viewId:(NSInteger)viewId;
- (void)setInfoBubblePositions:(NSArray *)positions;
- (void)moveCamera:(NSDictionary *)payload duration:(double)duration;
- (void)clearOverlays;
- (void)beginMarkerComposition:(NSInteger)generation icons:(NSArray *)icons;
- (void)appendMarkerComposition:(NSInteger)generation sequence:(NSInteger)sequence payload:(NSDictionary *)payload;
- (void)commitMarkerComposition:(NSInteger)generation;
- (void)updateMarker:(NSDictionary *)payload;
- (void)fitBounds:(NSDictionary *)bounds padding:(NSInteger)padding;
- (void)compositionCircles:(NSArray *)payload;
- (void)updateCircle:(NSDictionary *)payload;
- (void)compositionGroundImages:(NSArray *)payload;
- (void)updateGroundImage:(NSDictionary *)payload;
- (void)compositionPolygons:(NSArray *)payload;
- (void)updatePolygon:(NSDictionary *)payload;
- (void)compositionPolylines:(NSArray *)payload;
- (void)updatePolyline:(NSDictionary *)payload;
- (void)compositionRasterLayers:(NSArray *)payload;
- (void)updateRasterLayer:(NSDictionary *)payload;
- (void)upsertNativeMapExtension:(NSString *)extensionId type:(NSString *)type payload:(NSDictionary *)payload;
- (void)removeNativeMapExtension:(NSString *)extensionId;
@end

@interface MCMapLibreContainerView : UIView
@property(nonatomic, strong) MCMapLibreReactNativeView *mapView;
@property(nonatomic, copy) RCTDirectEventBlock onMapLoaded;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerCompositionBatchProcessed;
@property(nonatomic, copy) RCTDirectEventBlock onMapClick;
@property(nonatomic, copy) RCTDirectEventBlock onMapLongClick;
@property(nonatomic, copy) RCTDirectEventBlock onCameraMoveStart;
@property(nonatomic, copy) RCTDirectEventBlock onCameraMove;
@property(nonatomic, copy) RCTDirectEventBlock onCameraMoveEnd;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerClick;
@property(nonatomic, copy) RCTDirectEventBlock onCircleClick;
@property(nonatomic, copy) RCTDirectEventBlock onGroundImageClick;
@property(nonatomic, copy) RCTDirectEventBlock onPolylineClick;
@property(nonatomic, copy) RCTDirectEventBlock onPolygonClick;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerDragStart;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerDrag;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerDragEnd;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerAnimateStart;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerAnimateEnd;
@property(nonatomic, copy) RCTDirectEventBlock onMarkerScreenPositions;
@property(nonatomic, copy) RCTDirectEventBlock onInfoBubbleScreenPositions;
@property(nonatomic, copy) RCTDirectEventBlock onNativeMapExtensionEvent;
@end

@implementation MCMapLibreContainerView
- (instancetype)init {
  if ((self = [super init])) {
    Class cls = NSClassFromString(@"MCMapLibreReactNativeView");
    NSAssert(cls != Nil, @"MapConductorForMapLibre XCFramework is not linked");
    _mapView = [[cls alloc] initWithFrame:CGRectZero];
    [self addSubview:_mapView];
    __weak typeof(self) weakSelf = self;
    _mapView.eventHandler = ^(NSString *name, NSDictionary *body) { [weakSelf emit:name body:body]; };
  }
  return self;
}
- (void)layoutSubviews { [super layoutSubviews]; self.mapView.frame = self.bounds; }
- (void)emit:(NSString *)name body:(NSDictionary *)body {
  if ([name isEqualToString:@"mapLoaded"] && self.onMapLoaded) self.onMapLoaded(body);
  else if ([name isEqualToString:@"markerCompositionBatchProcessed"] && self.onMarkerCompositionBatchProcessed) self.onMarkerCompositionBatchProcessed(body);
  else if ([name isEqualToString:@"mapClick"] && self.onMapClick) self.onMapClick(body);
  else if ([name isEqualToString:@"mapLongClick"] && self.onMapLongClick) self.onMapLongClick(body);
  else if ([name isEqualToString:@"cameraMoveStart"] && self.onCameraMoveStart) self.onCameraMoveStart(body);
  else if ([name isEqualToString:@"cameraMove"] && self.onCameraMove) self.onCameraMove(body);
  else if ([name isEqualToString:@"cameraMoveEnd"] && self.onCameraMoveEnd) self.onCameraMoveEnd(body);
  else if ([name isEqualToString:@"markerClick"] && self.onMarkerClick) self.onMarkerClick(body);
  else if ([name isEqualToString:@"circleClick"] && self.onCircleClick) self.onCircleClick(body);
  else if ([name isEqualToString:@"groundImageClick"] && self.onGroundImageClick) self.onGroundImageClick(body);
  else if ([name isEqualToString:@"polylineClick"] && self.onPolylineClick) self.onPolylineClick(body);
  else if ([name isEqualToString:@"polygonClick"] && self.onPolygonClick) self.onPolygonClick(body);
  else if ([name isEqualToString:@"markerDragStart"] && self.onMarkerDragStart) self.onMarkerDragStart(body);
  else if ([name isEqualToString:@"markerDrag"] && self.onMarkerDrag) self.onMarkerDrag(body);
  else if ([name isEqualToString:@"markerDragEnd"] && self.onMarkerDragEnd) self.onMarkerDragEnd(body);
  else if ([name isEqualToString:@"markerAnimateStart"] && self.onMarkerAnimateStart) self.onMarkerAnimateStart(body);
  else if ([name isEqualToString:@"markerAnimateEnd"] && self.onMarkerAnimateEnd) self.onMarkerAnimateEnd(body);
  else if ([name isEqualToString:@"markerScreenPositions"] && self.onMarkerScreenPositions) self.onMarkerScreenPositions(body);
  else if ([name isEqualToString:@"infoBubbleScreenPositions"] && self.onInfoBubbleScreenPositions) self.onInfoBubbleScreenPositions(body);
  else if ([name isEqualToString:@"nativeMapExtensionEvent"] && self.onNativeMapExtensionEvent) self.onNativeMapExtensionEvent(body);
}
@end

// `RCT_EXPORT_MODULE` already generates a `+load` (to call `RCTRegisterModule`), so this uses a
// separate `__attribute__((constructor))` function rather than a second `+load` (which would be a
// duplicate-method compile error). `MapLibreMapView` isn't a Codegen'd Fabric component, so it
// needs this explicit legacy-interop registration to avoid "View config not found for component
// `MapLibreMapView`" under bridgeless New Architecture. Resolved dynamically (no
// `#import <React/RCTLegacyViewManagerInteropComponentView.h>`) so this pod doesn't take on
// React-RCTFabric's transitive C++/Yoga header dependency just to reach one registration call.
__attribute__((constructor))
static void MCMapLibreRegisterLegacyViewManagerInterop(void)
{
  Class cls = NSClassFromString(@"RCTLegacyViewManagerInteropComponentView");
  SEL selector = NSSelectorFromString(@"supportLegacyViewManagerWithName:");
  if (cls && [cls respondsToSelector:selector]) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
    [cls performSelector:selector withObject:@"MapLibreMapView"];
#pragma clang diagnostic pop
  }
}

@implementation MapConductorMapLibreViewManager
RCT_EXPORT_MODULE(MapLibreMapView)
- (UIView *)view { return [MCMapLibreContainerView new]; }
RCT_EXPORT_VIEW_PROPERTY(onMapLoaded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerCompositionBatchProcessed, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMapClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMapLongClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCameraMoveStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCameraMove, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCameraMoveEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onCircleClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onGroundImageClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPolylineClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPolygonClick, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerDragStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerDrag, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerDragEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerAnimateStart, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerAnimateEnd, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMarkerScreenPositions, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onInfoBubbleScreenPositions, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onNativeMapExtensionEvent, RCTDirectEventBlock)
RCT_CUSTOM_VIEW_PROPERTY(cameraPosition, NSDictionary, MCMapLibreContainerView) { if (json) [view.mapView setCameraPosition:json]; }
RCT_CUSTOM_VIEW_PROPERTY(mapDesignType, NSString, MCMapLibreContainerView) { [view.mapView setMapDesignType:json]; }
RCT_CUSTOM_VIEW_PROPERTY(markerTilingOptions, NSDictionary, MCMapLibreContainerView) { [view.mapView setMarkerTilingOptions:json viewId:view.reactTag.integerValue]; }
RCT_CUSTOM_VIEW_PROPERTY(infoBubblePositions, NSArray, MCMapLibreContainerView) { [view.mapView setInfoBubblePositions:json ?: @[]]; }
- (void)withView:(nonnull NSNumber *)tag block:(void (^)(MCMapLibreReactNativeView *))block {
  [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *manager, NSDictionary<NSNumber *, UIView *> *registry) {
    UIView *view = registry[tag];
    if ([view isKindOfClass:[MCMapLibreContainerView class]]) block(((MCMapLibreContainerView *)view).mapView);
  }];
}
RCT_EXPORT_METHOD(clearOverlays:(nonnull NSNumber *)tag) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v clearOverlays]; }]; }
RCT_EXPORT_METHOD(moveCamera:(nonnull NSNumber *)tag position:(nonnull NSDictionary *)p) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v moveCamera:p duration:0]; }]; }
RCT_EXPORT_METHOD(animateCamera:(nonnull NSNumber *)tag position:(nonnull NSDictionary *)p duration:(double)d) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v moveCamera:p duration:d]; }]; }
RCT_EXPORT_METHOD(beginMarkerComposition:(nonnull NSNumber *)tag generation:(NSInteger)g icons:(nonnull NSArray *)i) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v beginMarkerComposition:g icons:i]; }]; }
RCT_EXPORT_METHOD(appendMarkerComposition:(nonnull NSNumber *)tag generation:(NSInteger)g sequence:(NSInteger)s payload:(nonnull NSDictionary *)p) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v appendMarkerComposition:g sequence:s payload:p]; }]; }
RCT_EXPORT_METHOD(commitMarkerComposition:(nonnull NSNumber *)tag generation:(NSInteger)g) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v commitMarkerComposition:g]; }]; }
RCT_EXPORT_METHOD(updateMarker:(nonnull NSNumber *)tag payload:(nonnull NSDictionary *)p) { [self withView:tag block:^(MCMapLibreReactNativeView *v) { [v updateMarker:p]; }]; }

RCT_EXPORT_METHOD(fitBounds:(nonnull NSNumber *)tag bounds:(nonnull NSDictionary *)bounds padding:(NSInteger)padding) {
  [self withView:tag block:^(MCMapLibreReactNativeView *v) {
    if ([v respondsToSelector:@selector(fitBounds:padding:)]) [v fitBounds:bounds padding:padding];
  }];
}

#define MC_EXPORT_ARRAY_COMMAND(command) \
RCT_EXPORT_METHOD(command:(nonnull NSNumber *)tag payload:(nonnull NSArray *)payload) { \
  [self withView:tag block:^(MCMapLibreReactNativeView *v) { \
    if ([v respondsToSelector:@selector(command:)]) [v command:payload]; \
  }]; \
}

#define MC_EXPORT_OBJECT_COMMAND(command) \
RCT_EXPORT_METHOD(command:(nonnull NSNumber *)tag payload:(nonnull NSDictionary *)payload) { \
  [self withView:tag block:^(MCMapLibreReactNativeView *v) { \
    if ([v respondsToSelector:@selector(command:)]) [v command:payload]; \
  }]; \
}

MC_EXPORT_ARRAY_COMMAND(compositionCircles)
MC_EXPORT_OBJECT_COMMAND(updateCircle)
MC_EXPORT_ARRAY_COMMAND(compositionGroundImages)
MC_EXPORT_OBJECT_COMMAND(updateGroundImage)
MC_EXPORT_ARRAY_COMMAND(compositionPolygons)
MC_EXPORT_OBJECT_COMMAND(updatePolygon)
MC_EXPORT_ARRAY_COMMAND(compositionPolylines)
MC_EXPORT_OBJECT_COMMAND(updatePolyline)
MC_EXPORT_ARRAY_COMMAND(compositionRasterLayers)
MC_EXPORT_OBJECT_COMMAND(updateRasterLayer)

RCT_EXPORT_METHOD(upsertNativeMapExtension:(nonnull NSNumber *)tag extensionId:(nonnull NSString *)extensionId type:(nonnull NSString *)type payload:(nonnull NSDictionary *)payload) {
  [self withView:tag block:^(MCMapLibreReactNativeView *v) {
    if ([v respondsToSelector:@selector(upsertNativeMapExtension:type:payload:)]) {
      [v upsertNativeMapExtension:extensionId type:type payload:payload];
    }
  }];
}

RCT_EXPORT_METHOD(removeNativeMapExtension:(nonnull NSNumber *)tag extensionId:(nonnull NSString *)extensionId) {
  [self withView:tag block:^(MCMapLibreReactNativeView *v) {
    if ([v respondsToSelector:@selector(removeNativeMapExtension:)]) [v removeNativeMapExtension:extensionId];
  }];
}

#undef MC_EXPORT_ARRAY_COMMAND
#undef MC_EXPORT_OBJECT_COMMAND
@end

File: android/src/main/java/com/mapconductor/react/maplibre/MapConductorMapLibreViewManager.kt
package com.mapconductor.react.maplibre

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class MapConductorMapLibreViewManager : SimpleViewManager<MapLibreMapViewWrapper>() {
    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext): MapLibreMapViewWrapper =
        MapLibreMapViewWrapper(reactContext)

    override fun onAfterUpdateTransaction(view: MapLibreMapViewWrapper) {
        super.onAfterUpdateTransaction(view)
        view.initializeMapIfNeeded()
    }

    @ReactProp(name = "cameraPosition")
    fun setCameraPosition(
        view: MapLibreMapViewWrapper,
        cameraPosition: ReadableMap?,
    ) {
        view.setCameraPosition(cameraPosition)
    }

    @ReactProp(name = "mapDesignType")
    fun setMapDesignType(
        view: MapLibreMapViewWrapper,
        mapDesignType: String?,
    ) {
        view.setMapDesignType(mapDesignType)
    }

    @ReactProp(name = "infoBubblePositions")
    fun setInfoBubblePositions(
        view: MapLibreMapViewWrapper,
        positions: ReadableArray?,
    ) {
        view.setInfoBubblePositions(positions)
    }

    @ReactProp(name = "markerTilingOptions")
    fun setMarkerTilingOptions(
        view: MapLibreMapViewWrapper,
        options: ReadableMap?,
    ) {
        view.setMarkerTilingOptions(options)
    }

    override fun receiveCommand(
        root: MapLibreMapViewWrapper,
        commandId: String,
        args: ReadableArray?,
    ) {
        when (commandId) {
            "moveCamera" -> root.moveCamera(args?.getMap(0))
            "animateCamera" -> root.animateCamera(args?.getMap(0), args?.getInt(1) ?: 0)
            "fitBounds" -> root.fitBounds(args?.getMap(0), args?.getInt(1) ?: 0)
            "clearOverlays" -> root.clearOverlays()
            "compositionMarkers" -> root.compositionMarkers(args?.getMap(0))
            "beginMarkerComposition" ->
                root.beginMarkerComposition(
                    generation = args?.getInt(0) ?: return,
                    iconDictionary = if (args.size() > 1 && !args.isNull(1)) args.getArray(1) else null,
                )
            "appendMarkerComposition" ->
                root.appendMarkerComposition(
                    generation = args?.getInt(0) ?: return,
                    sequence = args.getInt(1),
                    payload = args.getMap(2),
                )
            "commitMarkerComposition" -> root.commitMarkerComposition(args?.getInt(0) ?: return)
            "updateMarker" -> root.updateMarker(args?.getMap(0))
            "compositionCircles" -> root.compositionCircles(args?.getArray(0))
            "updateCircle" -> root.updateCircle(args?.getMap(0))
            "compositionPolylines" -> root.compositionPolylines(args?.getArray(0))
            "updatePolyline" -> root.updatePolyline(args?.getMap(0))
            "compositionPolygons" -> root.compositionPolygons(args?.getArray(0))
            "updatePolygon" -> root.updatePolygon(args?.getMap(0))
            "compositionRasterLayers" -> root.compositionRasterLayers(args?.getArray(0))
            "updateRasterLayer" -> root.updateRasterLayer(args?.getMap(0))
            "compositionGroundImages" -> root.compositionGroundImages(args?.getArray(0))
            "updateGroundImage" -> root.updateGroundImage(args?.getMap(0))
            "upsertNativeMapExtension" ->
                root.upsertNativeMapExtension(
                    extensionId = args?.getString(0) ?: return,
                    type = args.getString(1) ?: return,
                    payload = args.getMap(2),
                )
            "removeNativeMapExtension" ->
                root.removeNativeMapExtension(args?.getString(0) ?: return)
        }
    }

    override fun onDropViewInstance(view: MapLibreMapViewWrapper) {
        view.onDropViewInstance()
        super.onDropViewInstance(view)
    }

    override fun getExportedCustomDirectEventTypeConstants(): MutableMap<String, Any> =
        mutableMapOf(
            "topMapLoaded" to mapOf("registrationName" to "onMapLoaded"),
            "topMarkerCompositionBatchProcessed" to
                mapOf("registrationName" to "onMarkerCompositionBatchProcessed"),
            "topMapClick" to mapOf("registrationName" to "onMapClick"),
            "topMapLongClick" to mapOf("registrationName" to "onMapLongClick"),
            "topCameraMoveStart" to mapOf("registrationName" to "onCameraMoveStart"),
            "topCameraMove" to mapOf("registrationName" to "onCameraMove"),
            "topCameraMoveEnd" to mapOf("registrationName" to "onCameraMoveEnd"),
            "topMarkerClick" to mapOf("registrationName" to "onMarkerClick"),
            "topCircleClick" to mapOf("registrationName" to "onCircleClick"),
            "topGroundImageClick" to mapOf("registrationName" to "onGroundImageClick"),
            "topPolylineClick" to mapOf("registrationName" to "onPolylineClick"),
            "topPolygonClick" to mapOf("registrationName" to "onPolygonClick"),
            "topMarkerDragStart" to mapOf("registrationName" to "onMarkerDragStart"),
            "topMarkerDrag" to mapOf("registrationName" to "onMarkerDrag"),
            "topMarkerDragEnd" to mapOf("registrationName" to "onMarkerDragEnd"),
            "topMarkerAnimateStart" to mapOf("registrationName" to "onMarkerAnimateStart"),
            "topMarkerAnimateEnd" to mapOf("registrationName" to "onMarkerAnimateEnd"),
            "topMarkerScreenPositions" to mapOf("registrationName" to "onMarkerScreenPositions"),
            "topInfoBubbleScreenPositions" to mapOf("registrationName" to "onInfoBubbleScreenPositions"),
            "topNativeMapExtensionEvent" to mapOf("registrationName" to "onNativeMapExtensionEvent"),
        )

    companion object {
        const val REACT_CLASS = "MapLibreMapView"
    }
}
